### PR TITLE
[DOCS] Add AST output

### LINKED ISSUE #1511: Document the AST output
The AST output from the compiler should be documented and we should follow our policy of breaking changes there as well.

### PR DESCRIPTION
**Pull request by ChrisChinchilla**
Closes https://github.com/ethereum/solidity/issues/1511

I wanted to get a PR open to discuss the best way to organise the information and fill in any gaps of nodes I have missed. There's not many standard ways to document AST outputs, so we have a lot of flexibility on how we can do it. I also think that an example with a diagram and the output would be useful.

I'm also not sure if the "child nodes" section for each node is helpful, or more reflective of the examples I have generated.

But generally feedback welcome on how we organise this information and what nodes and details might be missing.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [ ] Used meaningful commit messages

### CODE REVIEW

**Comment by chriseth** [comment]
Oh this is a horrible amount of work! :)
I think we should mainly document the output of ``solc --ast-compact-json``

File: docs/ast.rst [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+An array type that can be empty or contain an expression.
+
+Child nodes:
+
+- ``UserDefinedTypeName`` "VariableName"
+
+MappingTypeName
+===============
+
+.. TODO: Actually exist?
```
Comment (djudjuu) [inline docs/ast.rst:153]: no. ```libsolidity/ast/astjsonConverter.h``` is a handy source of truth

File: docs/ast.rst [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+================
+
+..  A literal function type. Its source form is "function (paramType1, paramType2) internal / external returns (retType1, retType2)"
+
+.. TODO
+
+UserDefinedTypeName "VariableName"
+==================================
+
+A user-defined type (i.e., a struct, contract, etc.).
```
Comment (djudjuu) [inline docs/ast.rst]: > (i.e., a struct, contract, etc.).

that sounds like they should be, but it  is not correct. 
e.g. the contract

``` contract C {struct A {uint x;}}``` will have a contractDefintion, structDefinition, and variableDeclaration.
UserDefinedTypeName is when you import some existing thing as X, e.g. in 

```contract A is B { B[] b;}```  => both B's will be a UserDefinedTypeName-Node

File: docs/ast.rst [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+========
+
+A continue statement that terminates execution of the statement in the current iteration of the current loop, and continues execution of the loop with the next iteration.
+
+Identifier "identifier"
+=======================
+
+An identifier, i.e., a reference to a declaration by name, such as a variable or function.
+
+.. TODO: Are fields just nodes?
```
Comment (djudjuu) [inline docs/ast.rst:259]:
```
{
  "argumentTypes" : null,
  "id" : 10,
  "name" : "j",
  "nodeType" : "Identifier",
  "overloadedDeclarations" : [],
  "referencedDeclaration" : 5,
  "src" : "95:1:0",
  "typeDescriptions" : 
    {
      "typeIdentifier" : "t_uint256",
      "typeString" : "uint256"
    }
}
```

**Comment by djudjuu** [comment]
> "I'm also not sure if the "child nodes" section for each node is helpful

Hmm, I think it would be helpful to have a brief explainer for all types of fields a node has. On the other hand, most of them have very obvious names (e.g. referenced Declarations), so that giving them an extra sentence might read weirdly and bloat the docs. Maybe you could do write out the children section for two nodes as an example. I would not refer to them as "children"-nodes though, as that what happens in the legacy-AST. I'd prefer "fields"

Also, there are a lot of references to ID which are not part of the AST. Those are pseudo-ASTNodes creating by parsing magicVariables, such as ```this, msg.sender, block.time```. So a note saying why those IDs do not exist would be in order somewhere. 

(atm, I am also working on assigning those dedicated negative IDs, I'll reference it here, once this is done. Just wanted to leave a reminder)

**Comment by djudjuu** [comment]
> generally feedback welcome on how we organise this information

one idea would be that we document it the same way as we are doing for the standard-json input and output: A json-object as an example with inline comments. A brief summary + the piece of code that generated the json + info which is not particular to any of the fields could go above.

**Comment by ChrisChinchilla** [comment]
@chriseth My only instruction was to look at `ast.h`, which is what I did… Hopefully this is all useful somewhere. And as this is a new file, I thought one large commit would be fine, not sure how else to do it really!

>I think we should mainly document the output of solc --ast-compact-json

What do you actually mean by this?

File: docs/ast.rst (hunk 1 above)
Comment (ChrisChinchilla) [reply docs/ast.rst:153]: @djudjuu does the file vary much from ast.h? That's where I was told to look.

File: docs/ast.rst (hunk 2 above)
Comment (ChrisChinchilla) [reply docs/ast.rst]: Ah, OK, yes. The comment in ast.h was a bit vague…

File: docs/ast.rst (hunk 3 above)
Comment (ChrisChinchilla) [reply docs/ast.rst:259]: @djudjuu What does this code example refer to?

**Comment by ChrisChinchilla** [comment]
@djudjuu 

https://github.com/ethereum/solidity/pull/7387#issuecomment-530745066

Do you mean like in `using-the-compiler.rst`? The main issue I find with documenting the AST output is that there are so many potential nodes, I have a series of example contracts I've been using to generate examples for me to work from which could cover a lot of cases, and I guess people can figure out the rest?

I'll look at some of your other comments too once we clear up some of these simpler ones.

**Comment by ChrisChinchilla** [comment]
@djudjuu Changed to fields, I agreed with that. I disagree with things being obvious though, we have had some feedback from people attempting to use the AST and not understanding it. Granted that was because there was _no_ documentation, but still, being overly clear doesn't hurt.

File: docs/ast.rst (hunk 3 above)
Comment (djudjuu) [reply docs/ast.rst:259]: > What does this code example refer to?

this is what a Identifier-JSON node looks like for ```uint j```... answering your "are fields just nodes"

**Review by djudjuu — COMMENTED** [review]
> Do you mean like in using-the-compiler.rst? 

jup. It would be consistent, but probably not the most readable. Still the best way to find out is to document one kind of node in both ways and compare. 

>  lot of cases, and I guess people can figure out the rest?
>  I disagree with things being obvious though, we have had some feedback from people attempting to use the AST and not understanding it.

Documentation should definitely cover all kinds of nodeTypes. Maybe some subfields can be left out, but probably it's easier to go for all.

File: docs/ast.rst (hunk 1 above)
Comment (djudjuu) [reply docs/ast.rst:153]: @ChrisChinchilla  It is where the ASTNode is used to generate the JSON-object.

Each ASTNode-class has a visit()-functino associated to it, where the attributes are processed and put into JSON-fields.

**Comment by djudjuu** [comment]
just leaving a note that the AST-output for inline assembly will change with #2419, once #7537 is merged.

**Comment by ChrisChinchilla** [comment]
Look into test suite inside `test/libsolidity/astjson/` for examples of input and output

File: docs/ast.rst [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+WhileStatement
+==============
+
+While loop statement.
+
+- BinaryOperation
+- Block
+
+IfStatement
+===========
```
Comment (Marenz) [inline docs/ast.rst:300]: For consistency, a blank line after this header would be good

File: docs/ast.rst [hunk 5]
```diff
@@ -0,0 +1,423 @@
...
+If statement with an optional "else" part. Note that "else if" is modeled by having a new if statement as the false (else) body.
+
+Fields:
+
+-  ``BinaryOperation``
```
Comment (Marenz) [inline docs/ast.rst:305]: Has one more space than the line below

**Comment by ChrisChinchilla** [comment]
Thanks @Marenz this is nowhere near finished, so formatting not finalised :)

**Comment by ChrisChinchilla** [comment]
OK, I've reworked this after feedback from @djudjuu a few weeks ago, and made it more like other examples in the "Using the compiler" section. A few remaining points:

- Even with a small code example, the AST output is really long, not sure what to do about that, open to ideas.
- Some of the definitions below are missing descriptions, as I wasn't sure what they represented, and couldn't find anything in the code base either.
- Another option is to move the explanations into code comments, which won't help with the length issue, but consolidates things

**Comment by Marenz** [comment]
I feel moving it into the comments is best...
Note that I had to do some more changes based on 060 to the output..

**Comment by ChrisChinchilla** [comment]
@Marenz Can you show me where I can find these changes?

**Comment by Marenz** [comment]
here you go: https://github.com/ethereum/solidity/pull/7925

**Comment by stackenbotten** [comment]
There was an error when running `chk_coding_style` for commit `108af00a98f1cbbe878425bd29f05f1d88faa135`:
```
Error: Trailing whitespace found:
 docs/ast.rst:452: // Defines the type of node, including: 
 docs/ast.rst:453: // SourceUnit, ContractDefinition, Block, FunctionDefinition, 

```
Please check that your changes are working as intended.

**Comment by ChrisChinchilla** [comment]
@chriseth Interviews with users summarised here - https://notes.ethereum.org/bEIZsmbbTLmPlP41fQIEqA?edit

**Comment by chriseth** [comment]
Closing for now.

**Pull request closed without merging**